plugins/linux-v4l2/v4l2-output.c

```diff
@@ -0,0 +1,160 @@
+#include <obs-module.h>
+#include <util/platform.h>
+#include <linux/videodev2.h>
+#include <sys/ioctl.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#define MAX_DEVICES 32
+
+struct virtualcam_data {
+	obs_output_t *output;
+	int device;
+	uint32_t frame_size;
+};
+
+static const char *virtualcam_name(void *unused)
+{
+	UNUSED_PARAMETER(unused);
+	return "Virtual Camera Output";
+}
+
+static void virtualcam_destroy(void *data)
+{
+	struct virtualcam_data *vcam = (struct virtualcam_data *)data;
+	close(vcam->device);
+	bfree(data);
+}
+
+static bool loopback_module_loaded()
+{
+	bool loaded = false;
+
+	FILE *fp = popen("lsmod | grep v4l2loopback", "r");
+	int status = pclose(fp);
+
+	if (WEXITSTATUS(status) == 2)
+		loaded = true;
+
+	return loaded;
+}
+
+static void loopback_module_load()
+{
+	system("pkexec modprobe v4l2loopback exclusive_caps=1 card_label='OBS Virtual Camera'");
+}
+
+static void *virtualcam_create(obs_data_t *settings, obs_output_t *output)
+{
+	struct virtualcam_data *vcam =
+		(struct virtualcam_data *)bzalloc(sizeof(*vcam));
+	vcam->output = output;
+
+	if (!loopback_module_loaded())
+		loopback_module_load();
+
+	UNUSED_PARAMETER(settings);
+	return vcam;
+}
+
+static bool try_connect(void *data, int device)
+{
+	struct virtualcam_data *vcam = (struct virtualcam_data *)data;
+	struct v4l2_format format;
+	struct v4l2_capability capability;
+	struct v4l2_streamparm parm;
+
+	uint32_t width = obs_output_get_width(vcam->output);
+	uint32_t height = obs_output_get_height(vcam->output);
+
+	vcam->frame_size = width * height * 2;
+
+	char new_device[16];
+	sprintf(new_device, "/dev/video%d", device);
+
+	vcam->device = open(new_device, O_RDWR);
+
+	if (vcam->device < 0)
+		return false;
+
+	if (ioctl(vcam->device, VIDIOC_QUERYCAP, &capability) < 0)
+		return false;
+
+	format.type = V4L2_BUF_TYPE_VIDEO_OUTPUT;
+
+	if (ioctl(vcam->device, VIDIOC_G_FMT, &format) < 0)
+		return false;
+
+	struct obs_video_info ovi;
+	obs_get_video_info(&ovi);
+
+	parm.type = V4L2_BUF_TYPE_VIDEO_OUTPUT;
+
+	parm.parm.output.timeperframe.numerator = ovi.fps_den;
+	parm.parm.output.timeperframe.denominator = ovi.fps_num;
```

Comment:
I think there are uninitialized values in the struct. Only timeperframe is set but the struct has more members:

```c
struct v4l2_outputparm {
	__u32		   capability;	 /*  Supported modes */
	__u32		   outputmode;	 /*  Current mode */
	struct v4l2_fract  timeperframe; /*  Time per frame in seconds */
	__u32		   extendedmode; /*  Driver-specific extensions */
	__u32              writebuffers; /*  # of buffers for write */
	__u32		   reserved[4];
};
```

The [documentation](https://www.kernel.org/doc/html/v4.9/media/uapi/v4l/vidioc-g-parm.html#id3) I found does not indicate that they can be left uninitialized.

Reply:
Done.